# OpenAPI Schema Validation Rules

## ü§ñ AI Agent Instructions

**WHEN**: Before generating any code that calls an API endpoint
**THEN**: Follow this validation process completely
**FAIL_IF**: Any validation step fails

---

## ‚ö†Ô∏è Critical Rule

```
IF generating_code_for_api_endpoint:
    MUST validate_against_openapi_schema() FIRST
    IF validation_fails:
        STOP and report_error()
    ELSE:
        PROCEED to generate_code()
```

**NEVER generate code before validating against the OpenAPI schema.**

---

## üìã Validation Process (Execute in Order)

### Step 1: Extract Endpoint Details

From `paths[endpoint][method]`:
- ‚úÖ Exact path (e.g., `/login`, `/items/{guid}`)
- ‚úÖ HTTP method (GET, POST, PUT, PATCH, DELETE)
- ‚úÖ Path parameters with types and formats
- ‚úÖ Query parameters with types and required flag
- ‚úÖ Header requirements

### Step 2: Resolve Request Schema

From `paths[endpoint][method].requestBody.content['application/json'].schema`:

1. **Extract the $ref**: `$ref: '#/components/schemas/LoginDetail'`
2. **Resolve to actual schema**: Navigate to `components.schemas.LoginDetail`
3. **Document all properties**:
   - Property names (exact camelCase)
   - Data types (string, integer, boolean, object, array)
   - Required fields (from `required` array)
   - Enums (from `enum` array)
   - Validation rules (patterns, min/max)

### Step 3: Resolve Response Schema

From `paths[endpoint][method].responses[200].content['application/json'].schema`:

1. **Extract the $ref**: `$ref: '#/components/schemas/LoginSuccess'`
2. **Resolve to actual schema**
3. **Identify critical fields**:
   - Session token field name
   - GUID fields
   - Status fields
   - Nesting structure

### Step 4: Validate Code Generation

**Decision Point**: Can I generate code now?

```
RUN validation_checklist():
    CHECK property_names_match_exactly(schema, generated_code)
    CHECK data_types_match_exactly(schema, generated_code)
    CHECK all_required_fields_included(schema, generated_code)
    CHECK no_extra_fields_added(generated_code, schema)
    CHECK parameter_locations_correct(schema, generated_code)
    
    IF all_checks_pass:
        RETURN true
    ELSE:
        RETURN false with failed_check_details

IF validation_checklist() == true:
    PROCEED to generate_code()
ELSE:
    STOP and fix_validation_errors()
```

**Validation Checklist** (MUST pass all):
- [ ] Property names match exactly (camelCase from schema)
- [ ] Data types match exactly (string‚ÜíString, int64‚ÜíLong, etc.)
- [ ] All required fields included in code
- [ ] No extra fields added beyond schema
- [ ] Parameter locations correct (path/query/header match schema)

## Example: Login Endpoint Validation

### OpenAPI Spec Analysis

```json
{
  "paths": {
    "/login": {
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/LoginDetail"
              }
            }
          }
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/LoginSuccess"
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "LoginDetail": {
        "type": "object",
        "properties": {
          "email": {"type": "string"},
          "password": {"type": "string"},
          "loginType": {
            "type": "string",
            "enum": ["EMAIL", "WEBTOKEN", "ACCESSTOKEN"]
          },
          "workspaceId": {"type": "integer", "format": "int64"}
        }
      },
      "LoginSuccess": {
        "type": "object",
        "properties": {
          "arenaSessionId": {"type": "string"},
          "workspaceId": {"type": "integer"},
          "workspaceName": {"type": "string"}
        }
      }
    }
  }
}
```

### Validated Code Generation

```java
// ‚úÖ CORRECT: Matches schema exactly
public class LoginDetail {
    private String email;           // ‚úÖ Correct type
    private String password;        // ‚úÖ Correct type
    private String loginType;       // ‚úÖ Correct type (enum)
    private Long workspaceId;       // ‚úÖ Correct type (int64 ‚Üí Long)
    
    // Getters and setters
}

public class LoginSuccess {
    private String arenaSessionId;  // ‚úÖ Correct name (camelCase)
    private Long workspaceId;       // ‚úÖ Correct type
    private String workspaceName;   // ‚úÖ Correct name
    
    // Getters and setters
}

// Login implementation
LoginDetail request = new LoginDetail();
request.setEmail(email);            // ‚úÖ Uses exact property name
request.setPassword(password);      // ‚úÖ Uses exact property name
request.setLoginType("EMAIL");      // ‚úÖ Uses valid enum value

String jsonBody = gson.toJson(request);
// Result: {"email":"...","password":"...","loginType":"EMAIL"}
```

### Common Mistakes (‚ùå)

```java
// ‚ùå WRONG: Property name doesn't match schema
public class LoginDetail {
    private String userName;  // ‚ùå Should be "email"
    private String pwd;       // ‚ùå Should be "password"
}

// ‚ùå WRONG: Wrong data type
public class LoginDetail {
    private String workspaceId;  // ‚ùå Should be Long/Integer
}

// ‚ùå WRONG: Extra fields not in schema
public class LoginDetail {
    private String email;
    private String password;
    private String sessionId;  // ‚ùå Not in LoginDetail schema
}

// ‚ùå WRONG: Wrong response field name
public class LoginSuccess {
    private String sessionId;  // ‚ùå Should be "arenaSessionId"
}
```

## Validation Checklist

Before generating code:

- [ ] Endpoint path is exact match
- [ ] HTTP method is correct
- [ ] Request schema $ref resolved
- [ ] All request properties match (name, type)
- [ ] Required fields identified
- [ ] Response schema $ref resolved
- [ ] Response parsing uses correct field names
- [ ] Enums use valid values
- [ ] Data type conversions are correct (int64 ‚Üí Long, etc.)
- [ ] No extra properties added
- [ ] No properties omitted

## Type Mapping Guidelines

| OpenAPI Type | Format | Java | TypeScript | Python |
|--------------|--------|------|------------|--------|
| string | - | String | string | str |
| string | date-time | String/LocalDateTime | string/Date | str/datetime |
| integer | int32 | Integer | number | int |
| integer | int64 | Long | number | int |
| number | float | Float | number | float |
| number | double | Double | number | float |
| boolean | - | Boolean | boolean | bool |
| array | - | List<T> | T[] | list |
| object | - | Object/Class | object/interface | dict/class |

## üîÄ AI Agent Decision Tree

```
START: User requests API operation
  ‚Üì
STEP 1: Locate endpoint in OpenAPI spec
  ‚Üì
FOUND? 
  ‚îú‚îÄ NO ‚Üí STOP: Report endpoint not found
  ‚îî‚îÄ YES ‚Üí Continue
  ‚Üì
STEP 2: Extract endpoint method (GET/POST/PUT/PATCH/DELETE)
  ‚Üì
STEP 3: Resolve request schema ($ref ‚Üí actual schema)
  ‚Üì
RESOLVED?
  ‚îú‚îÄ NO ‚Üí STOP: Report unresolved reference
  ‚îî‚îÄ YES ‚Üí Continue
  ‚Üì
STEP 4: Extract all properties with types and required flags
  ‚Üì
STEP 5: Resolve response schema ($ref ‚Üí actual schema)
  ‚Üì
STEP 6: Run validation checklist
  ‚Üì
ALL CHECKS PASS?
  ‚îú‚îÄ NO ‚Üí STOP: Report validation failures
  ‚îî‚îÄ YES ‚Üí Continue
  ‚Üì
STEP 7: Generate code with exact schema match
  ‚Üì
END: Code generation complete
```

---

## ‚úÖ Must Do (MANDATORY)

```
ALWAYS:
  - Read OpenAPI spec before generating code
  - Resolve all $ref references completely
  - Match property names exactly as in schema
  - Use correct data types from schema
  - Include all required fields
  - Validate enum values against schema
```

---

## ‚ùå Must Not Do (FORBIDDEN)

```
NEVER:
  - Guess property names
  - Add fields not in schema
  - Use wrong data types
  - Skip schema validation
  - Assume field names (always use exact camelCase from schema)
  - Generate code before reading spec
  - Ignore required fields
  - Use invalid enum values
```

---

## üéØ Quick Reference for AI Agents

**Before generating ANY API code, execute:**

```python
def validate_before_generating_code(endpoint, method):
    # Step 1: Get schema
    schema = openapi_spec['paths'][endpoint][method]
    
    # Step 2: Resolve request schema
    request_schema = resolve_ref(schema['requestBody']['content']['application/json']['schema'])
    
    # Step 3: Resolve response schema
    response_schema = resolve_ref(schema['responses']['200']['content']['application/json']['schema'])
    
    # Step 4: Validate
    assert all_properties_match(request_schema)
    assert all_types_correct(request_schema)
    assert all_required_fields_present(request_schema)
    assert no_extra_fields(request_schema)
    
    # Step 5: Only if all pass
    return True  # Now safe to generate code
```

