# AI Agent Quick Reference Card

## ğŸš€ Start Here

```
WHEN: User requests Arena PLM integration
THEN: Follow this flow
```

---

## ğŸ“ Step-by-Step Guide

### Step 1: Read Decision Tree
**File**: `/rules/README.md`
**Look for**: ğŸ¤– Quick Start for AI Agents

### Step 2: Match Pattern
```
IF CREATE or UPDATE:
    â†’ /rules/examples/03-metadata-driven.md
ELSE IF MULTIPLE operations:
    â†’ /rules/examples/02-series-operations.md
ELSE:
    â†’ /rules/examples/01-single-operation.md
```

### Step 3: Apply Patterns
```
ALWAYS apply:
  - /rules/patterns/session-management.md
  - /rules/patterns/three-layer-architecture.md
  - /rules/schemas/openapi-validation.md
```

### Step 4: Validate
```
RUN checklist from /rules/README.md
IF all_pass:
    PROCEED
ELSE:
    STOP
```

### Step 5: Generate Code
```
FOLLOW example structure
MATCH schema exactly
USE three layers
APPLY session pattern
```

---

## ğŸ¯ Quick Pattern Matcher

### Keywords Detection

```python
# Series Operations Pattern
keywords = ["then", "and then", "after that", "also", "followed by"]
if any(kw in user_request.lower() for kw in keywords):
    pattern = "Series Operations"

# Metadata-Driven Pattern
keywords = ["create", "update", "set", "change", "modify"]
if any(kw in user_request.lower() for kw in keywords):
    must_fetch_metadata = True
    pattern = "Metadata-Driven"

# Single Operation Pattern
operation_count = count_operations(user_request)
if operation_count == 1:
    pattern = "Single Operation"
```

---

## âœ… Mandatory Checks

### Before Generating Code

```
[ ] Read OpenAPI spec
[ ] Resolve $ref references
[ ] Extract properties and types
[ ] Validate against schema
[ ] Select correct pattern
[ ] Plan session management
[ ] Design three layers
```

### During Code Generation

```
[ ] Property names match exactly
[ ] Data types match exactly
[ ] All required fields present
[ ] No extra fields added
[ ] Session reuse implemented
[ ] All three layers generated
[ ] try-finally for cleanup
```

### After Code Generation

```
[ ] Code compiles
[ ] Matches example structure
[ ] Follows session pattern
[ ] Has all three layers
[ ] Error handling present
```

---

## ğŸ”€ Decision Flow

```
USER REQUEST
    â†“
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚ Is it CREATE/     â”‚
â”‚ UPDATE?           â”‚
â””â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”˜
     â”‚YES       â”‚NO
     â†“          â†“
[Example 3] â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
            â”‚ Multiple ops?      â”‚
            â””â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”˜
                 â”‚YES       â”‚NO
                 â†“          â†“
            [Example 2] [Example 1]
    â†“
APPLY:
  - Session pattern
  - Three-layer architecture
  - OpenAPI validation
    â†“
VALIDATE
    â†“
GENERATE
```

---

## ğŸ“‹ Rule Files Map

| File | Purpose | When to Read |
|------|---------|--------------|
| `/rules/README.md` | Navigation | Always first |
| `/rules/patterns/session-management.md` | Session lifecycle | Always |
| `/rules/patterns/three-layer-architecture.md` | Architecture | Always |
| `/rules/schemas/openapi-validation.md` | Validation | Before code gen |
| `/rules/examples/01-single-operation.md` | Single op | 1 operation |
| `/rules/examples/02-series-operations.md` | Multiple ops | 2+ operations |
| `/rules/examples/03-metadata-driven.md` | CREATE/UPDATE | Metadata needed |

---

## ğŸš¨ Critical Rules

```
RULE 1: Fetch metadata BEFORE CREATE/UPDATE
RULE 2: Login once, logout once
RULE 3: Generate all three layers
RULE 4: Match schema exactly
RULE 5: Use try-finally for cleanup
```

---

## âš ï¸ Common Mistakes

```
âŒ Logout between operations
âŒ Missing DomainApi layer
âŒ Wrong property names
âŒ Guess schema instead of reading
âŒ Skip metadata for CREATE/UPDATE
âŒ Multiple SessionManager instances
âŒ No try-finally cleanup
```

---

## ğŸ¯ Success Criteria

```
âœ… Code matches selected example
âœ… All three layers present
âœ… Session reused correctly
âœ… Schema validation passed
âœ… Metadata fetched if needed
âœ… Error handling included
âœ… Comments explain flow
```

---

## ğŸ’¡ Pro Tips

1. **Always read README first** - Has decision tree
2. **Match keywords** - Detect pattern from user words
3. **Validate early** - Check schema before generating
4. **Follow examples exactly** - They're tested and working
5. **Use checklists** - Don't skip validation steps

---

## ğŸ”— Quick Links

- **Start**: `/rules/README.md` â†’ ğŸ¤– Quick Start
- **Session**: `/rules/patterns/session-management.md` â†’ Decision Tree
- **Architecture**: `/rules/patterns/three-layer-architecture.md` â†’ Generation Steps
- **Validation**: `/rules/schemas/openapi-validation.md` â†’ Checklist
- **Examples**: `/rules/examples/*.md` â†’ Use This Pattern When

---

## ğŸ“ Emergency Checklist

If stuck, verify:
1. [ ] Read correct example for operation type?
2. [ ] Applied session management pattern?
3. [ ] Generated all three layers?
4. [ ] Validated against OpenAPI schema?
5. [ ] Followed example structure exactly?

If all yes â†’ Generate code
If any no â†’ Go back and fix

---

## ğŸ“ Learning Path

```
New to Arena PLM integration?

1. Read /rules/README.md (10 min)
2. Read /rules/examples/01-single-operation.md (5 min)
3. Read /rules/patterns/session-management.md (5 min)
4. Read /rules/patterns/three-layer-architecture.md (5 min)
5. Try generating simple search operation (10 min)

Total: 35 minutes to proficiency
```

---

## âœ… This Card Is Your Friend

**Bookmark this file**
**Reference it before every code generation**
**Follow the flow**
**Check the rules**
**Generate with confidence**

ğŸ¤– Happy coding! ğŸš€

